src/Binding/JSThread.cpp

```diff
@@ -413,56 +377,45 @@ static bool nidium_post_message(JSContext *cx, unsigned argc, JS::Value *vp)
 }
 
 
-static bool
-nidium_Thread_constructor(JSContext *cx, unsigned argc, JS::Value *vp)
+JSThread * JSThread::Constructor(JSContext *cx, JS::CallArgs &args,
+    JS::HandleObject obj)
 {
-    JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
-    JS::RootedObject ret(cx,
-                         JS_NewObjectForConstructor(cx, &Thread_class, args));
-
     JS::RootedScript parent(cx);
 
-    JSThread *nthread = new JSThread(ret, cx);
+    JSThread *nthread = new JSThread();
     JS::RootedFunction nfn(cx);
 
     if ((nfn = JS_ValueToFunction(cx, args[0])) == NULL
         || (nthread->m_JsFunction = JS_DecompileFunction(cx, nfn, 0)) == NULL) {
         printf("Failed to read Threaded function\n");
-        return true;
+        return nullptr;
     }
 
-    nthread->m_JsObject = ret;
+    nthread->m_JsObject = obj;
     nthread->m_Njs      = NJS;
 
     JS::AutoFilename af;
     JS::DescribeScriptedCaller(cx, &af, &nthread->m_CallerLineNo);
 
     nthread->m_CallerFileName = strdup(af.get());
 
-    JS_SetPrivate(ret, nthread);
-
-    args.rval().setObject(*ret);
-
-    JS_DefineFunctions(cx, ret, Thread_funcs);
-
-    return true;
+    return nthread;
 }
 
-static void Thread_Finalize(JSFreeOp *fop, JSObject *obj)
+JSFunctionSpec *JSThread::ListMethods()
 {
-    JSThread *nthread = static_cast<JSThread *>(JS_GetPrivate(obj));
+    static JSFunctionSpec funcs[] = {
+        CLASSMAPPER_FN(JSThread, start, 1),
+        JS_FS_END
+    };
 
-    if (nthread != NULL) {
-        delete nthread;
-    }
+    return funcs;
 }
 
-// }}}
-
-// {{{ Registration
-
-NIDIUM_JS_OBJECT_EXPOSE(Thread)
-
+void JSThread::RegisterObject(JSContext *cx)
+{
+    JSThread::ExposeClass<1>(cx, "Thread");
```

Comment:
Since the original JSClass has a `JSCLASS_HAS_RESERVED_SLOTS(1)`, it needs to be passed here as well.

```